### PR TITLE
Implemented Killough relative permeability hysteresis

### PR DESCRIPTION
**Pull request by francoishamon**
This PR implements table relative permeability hysteresis using the Killough model for use in the Northern Lights test case. 

The PR also slightly modifies `CompositionalMultiphaseBase.cpp` to allow for a history-dependent relperm, as we now need to do `relPerm.initializePhaseVolFractionState` and `relPerm.saveConvergedPhaseVolFractionState` to initialize and save the max/min historical saturations in the constitutive model.

Current status:
- [X] Non-wetting-phase hysteresis
- [X] Wetting-phase hysteresis

This PR adds:
- [X] Unit tests
- [X] An integrated test (actually, the PR modifies an existing one)

Rebaseline done in https://github.com/GEOSX/integratedTests/pull/189

### CODE REVIEW

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp [hunk 1]
```diff
@@ -545,29 +575,31 @@ void CompositionalMultiphaseBase::initializeFluidState( MeshLevel & mesh )
...
+    //
+    // Note:
+    // - This must be called after updateRelPermModel
+    // - This step depends phaseRelPerm
     updatePhaseMobility( subRegion, targetIndex );
-    updateCapPressureModel( subRegion, targetIndex );
-    // thermal conductivity is explicitly, so no update here
 
-    // 4.4 Finally, we initialize the thermal conductivity (which can depend on porosity and phase volume fraction)
+    // 4.4 Finally, we initialize the thermal conductivity
```
Comment (jafranc) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:589]: should be 4.5 ?

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp [hunk 2]
```diff
@@ -1510,15 +1550,12 @@ void CompositionalMultiphaseBase::implicitStepComplete( real64 const & time,
       capPressureMaterial.saveConvergedRockState( porosity, permeability );
     }
 
-    // Step 5: if the thermal option is on, send the converged porosity and phase volume fraction to the thermal conductivity model
+    // Step 6: if the thermal option is on, send the converged porosity and phase volume fraction to the thermal conductivity model
     // note: this is needed because the phaseVolFrac-weighted thermal conductivity treats phaseVolumeFraction explicitly for now
     if( m_thermalFlag )
     {
       arrayView2d< real64 const > const porosity = porousMaterial.getPorosity();
```
Comment (jafranc) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:1558]: should we repeat the note about pending thermal impl ?

**Review by jafranc — CHANGES_REQUESTED** [review]
Thanks @francoishamon for this nice PR, will be a great blueprint for capillary hysteresis.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp (hunk 1 above)
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp]: Yes, I made the change

**Review by jafranc — APPROVED** [review]

File: src/coreComponents/unitTests/constitutiveTests/testRelPerm.cpp [hunk 3]
```diff
@@ -573,14 +859,16 @@ TEST_F( RelPermTest, numericalDerivatives_VanGenuchtenBakerRelPermTwoPhase )
   real64 const eps = std::sqrt( std::numeric_limits< real64 >::epsilon() );
   real64 const tol = 1e-4;
 
-  real64 const start_sat = 0.3;
-  real64 const end_sat   = 0.7;
+  real64 const startSat = 0.3;
+  real64 const endSat   = 0.7;
   real64 const dS = 1e-1;
-  real64 const alpha = 0.4;
+
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/unitTests/constitutiveTests/testRelPerm.cpp:826]: I am using the `alpha` parameter for three-phase unit tests (to vary all the saturations at the same time), so it is not needed for two-phase flow.

File: src/coreComponents/unitTests/constitutiveTests/testRelPermHysteresis.cpp [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+  TableFunction & imbibitionTable_w = dynamicCast< TableFunction & >(
+    *functionManager.createChild( TableFunction::catalogName(), "imbibitionWater_swg" ) );
+  imbibitionTable_w.setTableCoordinates( coordinates_iw );
+  imbibitionTable_w.setTableValues( imbibitionValues_w );
+  imbibitionTable_w.reInitializeFunction();
+
+  imbibitionTable_w.setInterpolationMethod( TableFunction::InterpolationType::Linear );
+
+  TableFunction & imbibitionTable_g = dynamicCast< TableFunction & >(
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/unitTests/constitutiveTests/testRelPermHysteresis.cpp]: Good idea, I added a small utility function in the `constitutiveTestHelpers` to initialize the tables `testRelPerm.cpp` and `testRelPermHysteresis.cpp`.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:1559]: You are right. I added a note saying that the thermal implementation is not available yet.

File: src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.hpp [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+    // Step 1.c: find the new endpoint
+    // this is the saturation for the scanning curve endpoint
+    real64 const krwedAtScrt = drainageRelPermKernelWrapper.compute( &Scrt );
+    real64 const krwieStar = krwedAtScrt
+                             + deltak * pow( ( Smxd - Scrt ) / LvArray::math::max( minScriMinusScrd, ( Smxd - Smxi ) ), m_killoughCurvatureParam );
+
+    // Step 2: get the normalized value of saturation
+    real64 const ratio = ( Smxi - Swc ) / ( Scrt - Shy );
+    real64 const Snorm = Smxi - ( Scrt - S ) * ratio; // normalized saturation from equation 2.166
+    real64 const dSnorm_dS =  -ratio;
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.hpp]: Good catch! This reminded me that we should add the wetting-phase case in the unit test that verifies the derivatives of the hysteretic relperms.

File: src/coreComponents/unitTests/constitutiveTests/testRelPermHysteresis.cpp [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+  auto & imbibitionGasTableName = relPerm.getReference< string >( keys::imbibitionNonWettingRelPermTableNameString() );
+  imbibitionGasTableName = "imbibitionGas_swg";
+
+  relPerm.postProcessInputRecursive();
+  relPerm.initialize(); // to test all the checks
+  return relPerm;
+}
+
+class KilloughHysteresisTest : public ConstitutiveTestBase< TableRelativePermeabilityHysteresis >
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/unitTests/constitutiveTests/testRelPermHysteresis.cpp:268]: For now, I left this as is. It is done in a similar way in `testRelPerm.cpp` and `testCapillaryPressure.cpp`, so we should probably change them all at once (if we want to change that).

File: src/coreComponents/constitutive/relativePermeability/TableRelativePermeability.hpp [hunk 7]
```diff
@@ -123,6 +123,7 @@ class TableRelativePermeability : public RelativePermeabilityBase
 
   struct viewKeyStruct : RelativePermeabilityBase::viewKeyStruct
   {
+    static constexpr char const * relPermKernelWrappersString() { return "relPermWrappers"; }
```
Comment (klevzoff) [inline src/coreComponents/constitutive/relativePermeability/TableRelativePermeability.hpp:126]: Looks like you intended to use this string in wrapper registration, but forgot

File: src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.cpp [hunk 8]
```diff
@@ -0,0 +1,706 @@
...
+{
+  RelativePermeabilityBase::resizeFields( size, numPts );
+
+  integer const numPhases = numFluidPhases();
+
+  m_phaseMaxHistoricalVolFraction.resize( size, numPhases );
+  m_phaseMinHistoricalVolFraction.resize( size, numPhases );
+}
+
+void TableRelativePermeabilityHysteresis::initializePhaseVolFractionState( arrayView2d< real64 const, compflow::USD_PHASE > const & initialPhaseVolFraction ) const
```
Comment (klevzoff) [inline src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.cpp:617]: It seems to me `initializePhaseVolFractionState` and `saveConvergedPhaseVolFractionState` are doing the same thing, essentially. Assuming `m_phaseMinHistoricalVolFraction` is initialized to `1.0` and `m_phaseMaxHistoricalVolFraction` is initialized to `0.0`, both computations amount to
```c++
minHistS[ip] = min( minHistS[ip], currentS[ip] );
maxHistS[ip] = max( maxHistS[ip], currentS[ip] );
```
I would just keep `saveConvergedPhaseVolFractionState` and call it when solver is initializing the models with initial saturation.

**Review by klevzoff — APPROVED** [review]
Nice! I don't understand the model well enough to leave meaningful comments on the implementation, but I'm sure it's correct :)

File: src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.cpp (hunk 8 above)
Comment (klevzoff) [reply src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.cpp:617]: On a second thought, maybe this does not generalize well to all types of models. I'm struggling to come up with an example, but maybe are models for which "first time seeing state" and "update previously seen state" is a meaningful distinction that the model cannot figure out on its own, which necessitates having two functions in the interface.

**Pull request merged**